zebra-network/src/peer_set/set.rs

```diff
@@ -995,78 +1108,71 @@ where
         Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
 
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        self.poll_background_errors(cx)?;
+        // Update service and peer statuses.
+        //
+        // # Correctness
+        //
+        // All of the futures that receive a context from this method can wake the peer set buffer
+        // task. If there are no ready peers, and no new peers, network requests will pause until:
+        // - an unready peer becomes ready, or
+        // - a new peer arrives.
+
+        // Check for new peers, and register a task wakeup when the next new peers arrive. New peers
+        // can be infrequent if our connection slots are full, or we're connected to all
+        // available/useful peers.
+        let _poll_pending_or_ready: Poll<()> = self.poll_discover(cx)?;
+
+        // These tasks don't provide new peers or newly ready peers.
+        let _poll_pending: Poll<()> = self.poll_background_errors(cx)?;
+        let _poll_pending_or_ready: Poll<()> = self.inventory_registry.poll_inventory(cx)?;
+
+        // Check for newly ready peers, including newly added peers (which are added as unready).
+        // So it needs to run after `poll_discover()`. Registers a wakeup if there are any unready
+        // peers.
+        //
+        // Each connected peer should become ready within a few minutes, or timeout, close the
+        // connection, and release its connection slot.
+        //
+        // TODO: drop peers that overload us with inbound messages and never become ready (#7822)
+        let _poll_pending_or_ready: Poll<Option<()>> = self.poll_unready(cx)?;
```

Comment:
Could we add a correctness comment here like: "Must never register any task wakeups on this context after this call to `poll_unready()`"

Reply:
That's not how task wakeups work. (It's how I thought they worked when I initially opened this PR, but I was wrong.)

*Every* potential waking event gets a reference to the waker via the context, and *any* of those events can wake the task.

Reply:
> any of those events can wake the task

Like a select?

Reply:
Sort of, but much lower level. The context contains a waker, which gives the ability to wake the task.

Technically there is nothing stopping a future from storing a waker when it returns `Poll::Ready`, and in some cases in Zebra we do this in inner poll methods, in case the outer poll function returns `Pending` due to outer state or other inner poll methods.

> The typical life of a Waker is that it is constructed by an executor, wrapped in a [Context](https://docs.rs/futures/latest/futures/task/struct.Context.html), then passed to [Future::poll()](https://docs.rs/futures/latest/futures/future/trait.Future.html#tymethod.poll). Then, if the future chooses to return [Poll::Pending](https://docs.rs/futures/latest/futures/task/enum.Poll.html#variant.Pending), it must also store the waker somehow and call [Waker::wake()](https://docs.rs/futures/latest/futures/task/struct.Waker.html#method.wake) when the future should be polled again.

https://docs.rs/futures/latest/futures/task/struct.Waker.html

> When a future is not ready yet, poll returns Poll::Pending and stores a clone of the [Waker](https://doc.rust-lang.org/std/task/struct.Waker.html) copied from the current [Context](https://doc.rust-lang.org/std/task/struct.Context.html). This [Waker](https://doc.rust-lang.org/std/task/struct.Waker.html) is then woken once the future can make progress. For example, a future waiting for a socket to become readable would call .clone() on the [Waker](https://doc.rust-lang.org/std/task/struct.Waker.html) and store it. When a signal arrives elsewhere indicating that the socket is readable, [Waker::wake](https://doc.rust-lang.org/std/task/struct.Waker.html#method.wake) is called and the socket future’s task is awoken. Once a task has been woken up, it should attempt to poll the future again, which may or may not produce a final value.

https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll